### DISCUSSION TITLE
Payment workflow - Authorise, Confirm/Review then Capture

### DISCUSSION THREAD

**Discussion by engram-design** [craftcms/commerce#2328] [Ideas]
Just working with a client migrating from Magento to Commerce. They want to mimic their current checkout workflow, which is semi-common in the e-commerce world

It involves splitting the authorisation of payment gateways to the capturing. We would want users to fill out their address, shipping and payment information (enter credit card, or select PayPal), then proceed to a review/confirm step, then actually process the payment. I've seen this before on a few sites out there.

This would need to work for PayPal or Stripe. I suppose there should also be some sort of time limit from when the customer authorises their payment to triggering the payment.

Would this be possible?

**Comment by lukeholder** [comment]
I think this is a good feature.

I am just thinking about the implementation and the way this would work.

Option 1)

- Use the native gateway auth/capture functions.
- Have a setting per gateway instance called 'allow customer capture'.
- Add a public controller endpoint (with order customer/user authorization) to capture the authorized amount.

Questions on this approach:

1) If the customer captures the payment at the gateway level, how does this work for locations (like the USA) where the capture should only be done by the store at the time of shipping?

2) Working this way, the cart would become an order at the time of the initial authorization transaction. Would the store want a bunch of authorized abandoned orders? This might be confusing for a store owner who would then have the power to capture the transaction, even though the customer decided to abandon the cart. 


Option 2)

- The initial payment step could just be a 'create payment source' or 'add existing payment source to the order' step. This wouldn't turn the cart into an order.
- The second 'process payment' step would then be the regular gateway purchase request and could be a authorize or straight purchase. The capture would then be done by the store owner.


My feeling is that the second option is the preferred option.

There is nothing major stopping option 2 to be implemented right now as far as I know, as long as the gateway supports payment sources, which as far as I know only Stripe does at the moment.

Let me know your thoughts on the above.

**Comment by engram-design** [comment]
Ah, good thinking with the payment source idea in option 2, I think that could work. Only issue is that we'll want to use PayPal, which I suppose doesn't support payment sources. What's the best approach to add a FR for that, or is it even feasible?

**Comment by andris-sevcenko** [comment]
PayPal Rest gateway supports payment sources. I'm not sure about the availability of it in Australia, though.

**Comment by engram-design** [comment]
@andris-sevcenko I'll do some research! Otherwise, will continue on the PayPal gateway plugin repo.

**Comment by engram-design** [comment]
@andris-sevcenko Are payment sources possible (or feasible) with PayPal Express? As in, possible now, or possible with custom development, or just plain not possible?

**Comment by andris-sevcenko** [comment]
@engram-design isn't PayPal Express the gateway where you pay by your PayPal account? The payment source is your PayPal account in that case, no?

**Comment by engram-design** [comment]
For sure, just meant in the context of this question. Whether you can prompt the user to pay with PayPal, redirect off, approve purchase, head back to the site to review, then capture payment.

Maybe the only way to achieve this correctly is with @lukeholder's suggestion of a public controller action for this.
